Formula/p/phodav.rb

```diff
@@ -0,0 +1,25 @@
+class Phodav < Formula
+  desc "WebDav server implementation using libsoup (RFC 4918)"
+  homepage "https://gitlab.gnome.org/GNOME/phodav"
+  url "https://gitlab.gnome.org/GNOME/phodav.git", tag: "v3.0", revision: "d733fd853f0664ad8035b1b85604c62de0e97098"
+  license "LGPL-2.1-only"
+
+  depends_on "meson" => :build
+  depends_on "ninja" => :build
+  depends_on "pkg-config" => :build
+
+  depends_on "libsoup"
+  depends_on "xmlto"
+
+  uses_from_macos "libxml2"
+
+  def install
+    system "meson", "setup", "build", *std_meson_args
+    system "meson", "compile", "-C", "build"
+    system "meson", "install", "-C", "build"
+  end
+
+  test do
+    system "meson", "test", "-C", "build"
```

Comment:
Tests don't run in the source directory. And we need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

Reply:
@SMillerDev phodav is both an application and a library. It provides its own [unit tests](https://gitlab.gnome.org/GNOME/phodav/-/tree/d733fd853f0664ad8035b1b85604c62de0e97098/tests) that check _full_ functionality—ideally those are used as trying to roll one's own for a server-client application is tricky and would not be as complete.

One option would be to do the tests at `do install` and have `do test` return true—this catches any failures early rather than later. 

```ruby
  def install
    system "meson", "setup", "build", *std_meson_args
    system "meson", "compile", "-C", "build"
    system "meson", "test", "-C", "build"
    system "meson", "install", "-C", "build"
  end

  test do
    system "/usr/bin/true"
  end
```

Another option is to just check if a server can be cconstructed, though this will not be as thorough as the unit tests that come with phodav.

```ruby
  test do
    (testpath/"test.cpp").write <<~EOS
    #include <libphodav/phodav.h>
    #include <glib.h>
    int main() {
      GFile *root_dir = g_file_new_for_path("./phodav-virtual-root");
      GFile *real_dir = g_file_get_child(root_dir, "real");
      PhodavVirtualDir *root = phodav_virtual_dir_new_root();
      phodav_virtual_dir_root_set_real(root, "./phodav-virtual-root");
      PhodavVirtualDir *virtual_dir = phodav_virtual_dir_new_dir(root, "/virtual", NULL);
      phodav_virtual_dir_attach_real_child(virtual_dir, real_dir);
      PhodavServer *phodav = phodav_server_new_for_root_file(G_FILE(root));
      g_assert_nonnull(phodav);
      g_object_unref(virtual_dir);
      g_object_unref(real_dir);
      g_object_unref(root_dir);
      g_object_unref(root);
      SoupServer *server = phodav_server_get_soup_server(phodav);
      g_assert_nonnull(server);
      g_object_unref(phodav);
      return 0;
    }
EOS
    system ENV.cc, "test.cpp",
                   "-I#{Formula["glib"].include}/glib-2.0",
                   "-I#{Formula["glib"].lib}/glib-2.0/include",
                   "-I#{Formula["libsoup"].include}/libsoup-3.0",
                   "-I#{include}/libphodav-3.0",
                   "-L#{lib}",
                   "-lgio-2.0", "-lgobject-2.0", "-lglib-2.0",
                   "-lphodav-3.0",
                   "-o", "test"
    system "./test"
  end
```



Please advise.

Reply:
We want to test something that users will likely also do on their device, so not really running a testsuite. If constructing a server is normal usage, let's go with that.